packages/flutter/lib/src/widgets/system_context_menu.dart

```diff
@@ -0,0 +1,133 @@
+///
+/// Currently, only supports iOS and displays nothing on other platforms.
+///
+/// The context menu is the menu that appears, for example, when doing text
+/// selection. Flutter typically draws this menu itself, but this class deals
+/// with the platform-rendered context menu instead.
+///
+/// There can only be one system context menu visible at a time. Building this
+/// widget when the system context menu is already visible will hide the old one
+/// and display this one. A system context menu that is hidden is informed via
+/// [onSystemHide].
+///
+/// To check if the current device supports showing the system context menu,
+/// call [isSupported].
+///
+/// {@tool dartpad}
+/// This example shows how to create a [TextField] that uses the system context
+/// menu where supported and does not show a system notification when the user
+/// presses the "Paste" button.
+///
+/// ** See code in examples/api/lib/widgets/system_context_menu.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [SystemContextMenuController], which directly controls the hiding and
+///    showing of the system context menu.
+class SystemContextMenu extends StatefulWidget {
+  /// Creates an instance of [SystemContextMenu] that points to the given
+  /// [anchor].
+  const SystemContextMenu._({
+    super.key,
+    required this.anchor,
+    this.onSystemHide,
+  });
+
+  /// Creates an instance of [SystemContextMenu] for the field indicated by the
+  /// given [EditableTextState].
+  factory SystemContextMenu.editableText({
+    Key? key,
+    required EditableTextState editableTextState,
+  }) {
+    final (
+      startGlyphHeight: double startGlyphHeight,
+      endGlyphHeight: double endGlyphHeight,
+    ) = editableTextState.getGlyphHeights();
+    return SystemContextMenu._(
+      key: key,
+      anchor: TextSelectionToolbarAnchors.getSelectionRect(
+        editableTextState.renderEditable,
+        startGlyphHeight,
+        endGlyphHeight,
+        editableTextState.renderEditable.getEndpointsForSelection(
+          editableTextState.textEditingValue.selection,
+        ),
+      ),
+      onSystemHide: () {
+        editableTextState.hideToolbar();
+      },
+    );
+  }
+
+  /// The [Rect] that the context menu should point to.
+  final Rect anchor;
+
+  /// Called when the system hides this context menu.
+  ///
+  /// For example, tapping outside of the context menu typically causes the
+  /// system to hide the menu.
+  ///
+  /// This is not called when showing a new system context menu causes another
+  /// to be hidden.
+  final VoidCallback? onSystemHide;
+
+  /// Whether the current device supports showing the system context menu.
+  ///
+  /// Currently, this is only supported on newer versions of iOS.
+  static bool isSupported(BuildContext context) {
+    return defaultTargetPlatform == TargetPlatform.iOS
+        && (MediaQuery.maybeSupportsShowingSystemContextMenu(context) ?? false);
```

Comment:
does this check include the compile time opt-in flag?

Reply:
No, I haven't added a compile time opt-in flag. Currently users opt-in by returning SystemContextMenu from TextField.contextMenuBuilder. See the code block in the description at the top of this PR.

Reply:
If we use `SystemContextMenu`, then I don't think we need the compile time flag anymore since they explicitly opt-in by changing their code.

Reply:
Let's discuss the pros and cons of new API vs existing API. 

A few strawman idea: 

**New API**: 
- [pro] developers can be explicit about which menu to use for different text fields
- [con] Larger API surface - they need to write extra logic 
- [con] developers need to migrate their code to get secure paste. 

**Use existing API**: 
- [pro] no migration needed. Developers get secure paste out of box for free
- [pro] smaller API size. less to learn
- [con] less flexible - cannot use different menu for different text fields

Reply:
For the option "use existing API", you also mean that it would be on by default? I want to separate my thoughts on those two ideas:

### Build this feature into existing widgets
We could either:

 1. Build this into [TextField._defaultContextMenuBuilder](https://github.com/flutter/flutter/blob/500ed0b6f639d734b243e61e513593c88042b68f/packages/flutter/lib/src/material/text_field.dart#L838C18-L838C29) (and same for CupertinoTextField).
 2. Build this into [AdaptiveTextSelectionToolbar.build](https://github.com/flutter/flutter/blob/500ed0b6f639d734b243e61e513593c88042b68f/packages/flutter/lib/src/material/adaptive_text_selection_toolbar.dart#L302) (same for CupertinoAdaptiveTextSelectionToolbar).

And create a parameter on the relevant widget to turn this on/off.

I think we'd still have to keep SystemContextMenu and SystemContextMenuController public in both cases. I also think it's cleaner to specify this using widget composition rather than adding a new parameter. But that's assuming that it's off by default.

### Use system context menus by default (where supported)
We could always do this later without changing any APIs, but my main hesitation for not doing it now is just fear of this big change. I guess I have a subconscious hope that Apple will fix this and we can return to Flutter-drawn menus at some point. But maybe I'm naive and we should just do it. The reality right now is that I expect most users will want to avoid the secure paste bug by using the native menu.